ipykernel/iostream.py

```diff
@@ -105,87 +109,183 @@ def _check_mp_mode(self):
+        self._io_loop = pub_thread.io_loop
+        self._new_buffer()
 
-                if hasattr(self.pub_socket, 'flush'):
-                    # socket itself has flush (presumably ZMQStream)
-                    self.pub_socket.flush()
+    def _is_master_process(self):
+        return os.getpid() == self._master_pid
+
+    def _check_mp_mode(self):
+        """check for forks, and switch to zmq pipeline if necessary"""
+        if self._is_master_process():
+            return MASTER
         else:
-            with self._pipe_out_lock:
-                string = self._flush_buffer()
-                tracker = self._pipe_out.send_multipart([
-                    self._pipe_uuid,
-                    string.encode(self.encoding, 'replace'),
-                ], copy=False, track=True)
-                try:
-                    tracker.wait(1)
-                except:
-                    pass
+            return CHILD
 
+    def set_parent(self, parent):
+        self.parent_header = extract_header(parent)
+
+    def close(self):
+        self.pub_thread = None
+
+    @property
+    def closed(self):
+        return self.pub_thread is None
+
+    def _schedule_flush(self):
+        """schedule a flush in the IO thread
+        
+        call this on write, to indicate that flush should be called soon.
+        """
+        with self._flush_lock:
+            if self._flush_timeout is not None:
+                return
+            # non-None to avoid races while waiting for _schedule
+            self._flush_timeout = 'placeholder'
+        
+        # add_timeout has to be handed to the io thread with add_callback
+        def _schedule_in_thread():
+            # FIXME: call_later is a little nicer, if we can require tornado ≥ 4
+            self._flush_timeout = self._io_loop.add_timeout(
+                self._io_loop.time() + self.flush_interval, self._flush)
+        self._io_loop.add_callback(_schedule_in_thread)
+
+    def flush(self):
+        """trigger actual zmq send
+        
+        send will happen in the background thread
+        """
+        if self.pub_thread.thread.is_alive():
+            self._io_loop.add_callback(self._flush)
+            # wait for flush to actually get through:
+            evt = threading.Event()
+            self._io_loop.add_callback(evt.set)
+            evt.wait()
+        else:
+            self._flush()
+    
+    def _flush(self):
+        """This is where the actual send happens.
+        
+        _flush should generally be called in the IO thread,
+        unless the thread has been destroyed (e.g. forked subprocess).
+        """
+        with self._flush_lock:
+            self._flush_timeout = None
+            data = self._flush_buffer()
+        if data:
+            # FIXME: this disables Session's fork-safe check,
+            # since pub_thread is itself fork-safe.
+            # There should be a better way to do this.
+            self.session.pid = os.getpid()
```

Comment:
Won't this disable the fork check for everything using that same session? Could that cause problems if other things then try to use non-fork-safe bits of Session?